IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometryResource/Functions/IfcGradient/Expression.txt

```diff
@@ -1,2 +1,36 @@
-(GradientCurve : IfcGradientCurve) : IfcLengthMeasure;
-  RETURN(1);
+(GradientCurve : IfcGradientCurve, Position : IfcCurveMeasureSelect) : IfcLengthMeasure;
+    
+  REPEAT i := 1 TO (SIZEOF(GradientCurve.Segments) - 1);
+    Segment : IfcCurveSegment := GradientCurve.Segments[i];
+    NextSegment : IfcCurveSegment := GradientCurve.Segments[i + 1];
+    Location : IfcCartesianPoint := NextSegment\IfcCurveSegment.Placement\IfcAxis2Placement2D.Location\IfcCartesianPoint;
+    SegmentPlacement : IfcAxis2Placement2D := Segment.Placement\IfcAxis2Placement2D;
+	
+    IF (Location[1] > Position)
+      CASE TRUE OF
+	  
+        'IFCCIRCLE' IN TYPEOF(Segment.BaseCurve) :
+          BEGIN
+            RETURN SegmentPlacement.Location\IfcCartesianPoint[2] + Segment.BaseCurve.Radius * (COS(ACOS(SegmentPlacement.RefDirection[1]) / 2 - Segment.SegmentLength) - (Position - SegmentPlacement.Location\IfcCartesianPoint[1]));
+          END;
+		  
+        'IFCCLOTHOID' IN TYPEOF(Segment.BaseCurve) :
+          BEGIN			
+            RETURN SegmentPlacement.Location\IfcCartesianPoint[2] + 1/3 * Segment.SegmentLength ** 3 + 1/(3 * IfcFactorial(7)) * Segment.SegmentLength ** 7 + 1/(5 * IfcFactorial(11)) * Segment.SegmentLength ** 11;
+          END;		
+	  
+          'IFCLINE' IN TYPEOF(Segment.BaseCurve) :
+            BEGIN
+              RETURN SegmentPlacement.Location\IfcCartesianPoint[2] + Segment.BaseCurve\IfcLine.Pnt[2] + Segment.BaseCurve\IfcLine.Dir.Orientation[2] * Segment.BaseCurve\IfcLine.Dir.Magnitude * Segment.SegmentLength;
```

Comment:
```suggestion
              RETURN SegmentPlacement.Location\IfcCartesianPoint.Coordinates[2] + Segment.BaseCurve\IfcLine.Dir.Orientation.DirectionRatios[2] * Segment.SegmentLength;
```

1. Similarly for all formulas where `IfcCartesianPoint` and `IfcDirection` are used.

2. I also think that `Magnitude` should be ignored here: we are interested in the direction, not the magnitude in that direction - that is provided by `SegmentLength`.

3. Reasoning for the deletion of `Segment.BaseCurve\IfcLine.Pnt[2]`: I think that the start point of the segment and the start point of the referenced curve are to coincide by definition (sth similar is missing from all other Cases).

Reply:
Ad 2.: Disregard the comment. I forgot about the `IfcLine` parametrization. This needs some more thought ...